vdev/src/testing/integration.rs

```diff
@@ -34,8 +40,15 @@ impl IntegrationTest {
         };
         let network_name = format!("vector-integration-tests-{integration}");
         let compose = Compose::new(test_dir, env_config.clone(), network_name.clone())?;
+
+        let full_runner_avail = full_runner_available()?;
+
+        let runner_name = (!full_runner_avail)
+            .then(|| (!all).then(|| integration.clone()))
+            .flatten();
```

Comment:
😞  I agree this logic could be a little more clear or at least documented.

The intended logic is:
- the `all` (or `build_all`) flag indicates we want to build with the all-features flag
- the runner name is either None or Some(<integration>). 
    - it is None if we already have a full runner available , or if `all` is true
    - it is Some(integration) if there is no full runner available and `all` is false (was trying to save local developers from having to re compile for a specific integration, if they already had a cached full image)

Looking at it now, it strikes me that this logic would probably be better to move to the `IntegrationTestRunner` struct though, passing the `all` argument and the integration name to the `::new()` function.

> That is, if `all` is true, should we ever check if the full runner is available and instead just build it?

So the building of the runner still occurs whether `all` is true or not. It's just that if `all` was specified, we use the runner image name that is not integration specific ( "vector-test-runner-<rust version>") , and when it builds, it uses  any cached steps it can, which for the integration tests would be all of them, and for local developers it would compile the difference.

But yes good catch, to your point, this should be checking the all flag first, followed by the runner available.

> I'd rather the behavior be explicit with some kind of "build all" flag

With that prior explanation now, I want to check if this is still a concern for you- I do think changing the name to `build_all` is better.
But do you think the edit to line 66 (essentially forcing the compilation with all features flag if we detected that the image was available), is being "too helpful" ?

Reply:
Yes, I do think it might be "too helpful". Having said that, we have `vdev` do the "helpful" thing by starting and stopping the integration environment when it wasn't previously started, so it's not a hard rule. I'm just thinking that, once that "all" image is built, there is no way to get back to only building the individual tests other than by deleting the image. That goes beyond just helping to set up running state.

Reply:
yeah that is a good point.

Possible solutions:

1. only use the full featured runner if the `build_all` is specified (dont be [un]helpful)
2. in the `stop` have a flag that when specified removes the full featured runner image if its there.
3. in the `stop` always remove the full featured runner image if its there.

Option 1 keeps it simple and all the usages available.

Option 2 is consistency. Users of the `build_all` option would be presumed to know what they are doing and know to use the relevant flag in the `stop` as needed.

Option 3 is kind of a mixed bag. And actually, because users sometimes run the `stop` to cleanup stray services or if they cancelled a run midway , that could remove the full runner when they didn't want to.

I'm def leaning to option 1 (which I think was what you had in mind) , unless you have other thoughts.

Reply:
I would agree with your leaning.

Reply:
addressed in [2d80832](https://github.com/vectordotdev/vector/pull/17724/commits/2d808324369fbe6dfe7aa5f8e72bf7a43b217c50)